crates/agglayer-node/src/rpc/rpc_middleware/logging_timeout.rs

```diff
@@ -0,0 +1,97 @@
+//! A logging timeout layer.
+
+use std::{borrow::Cow, future::Future, time::Duration};
+
+use futures::TryFutureExt;
+use jsonrpsee::{
+    server::middleware::rpc::RpcServiceT,
+    types::{ErrorObject, Id, Request},
+    MethodResponse,
+};
+use tracing::warn;
+
+/// Error code to return when the response time is too long.
+pub const TIMEOUT_ERROR_CODE: i32 = -32001;
+
+/// A layer that applies a timeout on a request and issues a log entry if the
+/// timeout expires before the request is completed.
+#[derive(Clone, Debug)]
+pub struct LoggingTimeoutLayer {
+    timeout: Duration,
+}
+
+impl LoggingTimeoutLayer {
+    pub fn new(timeout: Duration) -> Self {
+        Self { timeout }
+    }
+}
+
+impl<S> tower::Layer<S> for LoggingTimeoutLayer {
+    type Service = LoggingTimeoutService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        LoggingTimeoutService::new(self.timeout, inner)
+    }
+}
+
+pub struct LoggingTimeoutService<S> {
+    inner: S,
+    timeout: Duration,
+}
+
+impl<S> LoggingTimeoutService<S> {
+    pub fn new(timeout: Duration, inner: S) -> Self {
+        Self { inner, timeout }
+    }
+}
+
+impl<'a, S: RpcServiceT<'a>> RpcServiceT<'a> for LoggingTimeoutService<S> {
+    type Future = LoggingTimeoutFuture<'a, S::Future>;
+
+    fn call(&self, request: Request<'a>) -> Self::Future {
+        LoggingTimeoutFuture {
+            timeout: self.timeout,
+            method: request.method.clone(),
+            request_id: request.id.clone(),
+            inner: self.inner.call(request),
+        }
+    }
+}
+
+#[pin_project::pin_project]
+pub struct LoggingTimeoutFuture<'a, F> {
+    // Timeout duration.
+    timeout: Duration,
+
+    // Method information.
+    method: Cow<'a, str>,
+    request_id: Id<'a>,
```

Comment:
```suggestion
     /// Unique ID for the request.
    request_id: Id<'a>,
```

Reply:
Will package the two fields into a helper struct. It's the same between the two middlewares.